### ISSUE TITLE
Error with non cached JSON/Hash value in v3.0.0

### ISSUE THREAD

**Issue by jonian** [shrinerb/shrine#420]
Hi, thanks for this gem, it is amazing!

After updating to version 3.0.0, I'm getting an error when trying to update an existing record using JSON/Hash as value.

In versions before 3.0, the assign method checks for cached files:

```ruby
def assign(value, **options)
  if value.is_a?(String)
    return if value == "" || !cached?(uploaded_file(value))
    assign_cached(uploaded_file(value))
  else
    uploaded_file = cache!(value, action: :cache, **options) if value
    set(uploaded_file)
  end
end
```

While in version 3.0, the assign method treats String/Hash values always as cached files:

```ruby
def attach_cached(value, **options)
  if value.is_a?(String) || value.is_a?(Hash)
    change(cached(value, **options), **options)
  else
    attach(value, storage: cache_key, action: :cache, **options)
  end
end
```

```
/home/jonian/.gem/ruby/2.6.0/gems/shrine-3.0.0/lib/shrine/attacher.rb:355:in `cached': expected cached file, got "{\"id\":\"534e009d3701166238fe95557b737661.jpg\",\"storage\":\"store\",\"metadata\":{\"size\":44399,\"color\":\"#091C26\",\"width\":1920,\"height\":1080,\"palette\":[\"#091C26\",\"#2C4B56\",\"#1E3A46\",\"#4D6A74\"],\"filename\":\"Rain.jpg\",\"mime_type\":\"image/jpeg\"}}" (Shrine::Error)
```

**Comment by janko** [comment]
This is actually desired behaviour. Previously, assigning a non-cached file would be silently ignored, and now an explicit exception is being raised telling you it's not allowed. Is there any reason why the old behaviour would be desired?

Yes, it is a breaking change, but I wanted to take the opportunity of the major version bump to make it. Though I realized I forgot to list it in the release notes and changelog.

**Comment by Sexual** [comment]
@janko Is there a way to globally patch this? When using an SPA, this change breaks editing records with an already attached file.

**Comment by jonian** [comment]
@Sexual, you can use this monkey patch:

```ruby
class Shrine
  class Attacher
    module InstanceMethods
      alias attach_cached_original attach_cached

      def attach_cached(value, **options)
        if value.is_a?(String) || value.is_a?(Hash)
          file = uploaded_file(value)
          return unless cached?(file)
        end

        attach_cached_original(value, **options)
      end
    end
  end
end
```

**Comment by jrochkind** [comment]
You are saying you want/need behavior where instead of getting an error, the assignment simply doesn't work without an error, it doesn't do anything at all?

Are you sure that isn't a bug in your app, where you app thinks it's attaching a file but fails to do so?  Why do you want this behavior? How is it related to Single Page Applications?

I ask because maybe there's something we don't understand and some reason shrine's behavior doesn't work for you; maybe our understanding of the situation isn't accurate. Monkey patching the old behavior in is probably not a great idea, and could open you up to future bugs.

**Comment by Sexual** [comment]
@jrochkind We're using presigned uploads and therefore update the ActiveRecord model via a standard `Model.update` with the permitted params. For example:

The `Book` model has a `image` Shrine attribute. The API would return the following response:

```
{ 
   "id":"6",
   "image":{ 
      "id":"eb9f284230d2b8fa71c6b4b6ee3a40c394da0e23a4b990161636e127212984f3",
      "storage":"books",
      "metadata":{ 
         "size":2198,
         "filename":"image.png",
         "mime_type":"image/png"
      }
   }
}
```

When updating the record, but not the `image`, the frontend automatically "stringifies" the `image` parameter, which would work perfectly on Shrine 2, but now raises an error.

**Comment by Sexual** [comment]
@jonian How would you recommend including this patch in a Rails app. We'd like to avoid forking the gem.

**Comment by janko** [comment]
@Sexual You can put what @jonian suggested in your Shrine initializer, but you can define it on the `Shrine::Attacher` class directly to be able to use `super`:

```rb
class Shrine::Attacher
  def attach_cached(value, **options)
    if value.is_a?(String) || value.is_a?(Hash)
      file = uploaded_file(value)
      return unless cached?(file)
    end

    super
  end
end
```

This discussion actually gave me a bit to think about. I believe the problem is not that we're raising an exception for non-cached files, but that we're raising an exception for the *same* file. In other words, assigning a different file could remain being disallowed if it's not cached, but assigning the same file *could* be a no-op (whether it's cached or stored).

```rb
book.image #=> #<Shrine::UploadedFile id="foo" storage=:store>
book.image = '{"id":"bar","storage":"store",...}' # fails
book.image = '{"id":"foo","storage":"store",...}' # no-op
```

However, it would require additional logic, and I don't think this is a strong enough use case. It sounds like you could work around it by having the API *not* return Shrine's attachment data in the `image` attribute of the Book resource. I believe the frontend will only ever care about the attachment URL and metadata, which can be exposed separately. That way `image` attribute would remain blank on updates.

**Comment by Sexual** [comment]
@janko I strongly disagree with your opinion that this is not a strong enough use case to add the additional logic. It seems to directly go against the standard where if something is not being updated, the parameter is ignored. 

It would massively complicate stuff for us/anyone else to add additional logic on both the backend and frontend to discard the file parameter if it has not changed or manually add additional attributes for the `image` metadata/URL, but then also have to determine when to use the real `image` attribute.

Whilst I do understand why Shrine has gone this direction with raising an error, I believe it has unintentional consequences and might be worth at least adding a config option to enable/disable this if the default won't be changed.

**Comment by janko** [comment]
@Sexual Ok, that's fair, then I want you to help me understand. At the moment I don't know why it's useful to return raw attachment data from the API to the frontend in the first place. What does the frontend do with this data in your application?

What I imagine is that the frontend needs the URL to the uploaded files, which cannot be generally generated from the raw data on the frontend, so the API needs to return it as well. Does this URL then get returned also in the `image` attribute as part of that data, or do you return it separately?

> It seems to directly go against the standard where if something is not being updated, the parameter is ignored.

Shrine's attachment attribute is not exactly a regular parameter, because the value you put in is not the value you get out; `record.attachment=` accepts a raw IO object or Hash/String data, and `record.attachment` returns a `Shrine::UploadedFile` object. So I wouldn't automatically say it should go by the same standards as other parameters.

> Whilst I do understand why Shrine has gone this direction with raising an error, I believe it has unintentional consequences and might be worth at least adding a config option to enable/disable this if the default won't be changed.

Adding support for your use case wouldn't be difficult by itself, but these small additions add up. That's why I initially wasn't in favor of it, but I'm open to it now.

**Comment by Sexual** [comment]
@janko We display all the raw data such as file name, size, extension. We have used both a separate attribute in our serialise for `image_url` as well as adding the `url` attribute to `image`.

**Comment by janko** [comment]
This functionality has been merged into base after all (no additional plugin needs to be loaded), and it's available now in Shrine 3.2.0.
